### PR TITLE
raise error if pytest ret code > 1

### PR DESCRIPTION
**Pull request by d-chambers**
### What does this PR do?
Raises an Error if pytest fails to run when using `obspy-rutests`, thereby avoiding submitting an empty test report. 

### Why was it initiated?  Any relevant Issues?
See #2953

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just add the "build_docs" tag to this PR.
- [ ] If all tests including network modules (e.g. `clients.fdsn`) should be tested for the PR,
      just add the "test_network" tag to this PR.
- [x] All tests still pass.
- [x] Any new features or fixed regressions are covered via new tests.
- [x] Any new or changed features are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .
- [x] If the changes affect any plotting functions you have checked that the plots
      from all the CI builds look correct. Add the "upload_plots" tag so that plotting 
      outputs are attached as artifacts. 
- [x] Add the "ready for review" tag when you are ready for the PR to be reviewed.

### CODE REVIEW

File: obspy/scripts/runtests.py [hunk 1]
```diff
@@ -109,10 +109,19 @@ def main():
     os.chdir(base_obspy_path)
     try:
         status = pytest.main(plugins=[plugin])
+    except Exception as e:
```
Comment (trichter) [inline obspy/scripts/runtests.py:112]: Catching the exception and reraising it. I think this has no effect.

File: obspy/scripts/runtests.py [hunk 2]
```diff
@@ -109,10 +109,19 @@ def main():
...
+    else:
+        if int(status) > 1:
+            msg = (
+                'A failure occurred while running ObsPys test suite. Please '
+                'check the output carefully and try again.'
+                   )
+            raise ValueError(msg)
+        upload_json_report(report=report, data=plugin.report)
+        sys.exit(status)
     finally:
```
Comment (trichter) [inline obspy/scripts/runtests.py:123]: I'd move the finally block up again and the new code out of the try block.

File: obspy/scripts/runtests.py [hunk 3]
```diff
@@ -109,10 +109,19 @@ def main():
...
         status = pytest.main(plugins=[plugin])
+    except Exception as e:
+        raise e
+    else:
+        if int(status) > 1:
+            msg = (
+                'A failure occurred while running ObsPys test suite. Please '
+                'check the output carefully and try again.'
+                   )
+            raise ValueError(msg)
```
Comment (trichter) [inline obspy/scripts/runtests.py:120]: Maybe just `sys.exit(msg)` and also reporting the pytest return code in msg?

File: obspy/scripts/runtests.py (hunk 1 above)
Comment (d-chambers) [reply obspy/scripts/runtests.py:112]: It was just required to use an `else` clause

Comment (trichter) [reply obspy/scripts/runtests.py:112]: OK, I see, but here you are basically catching an exception inside pytest. I have not encountered such a case.

Comment (d-chambers) [reply obspy/scripts/runtests.py:112]: Ya, it was required syntactically to have an else clause. I will just refactor though so we don't need either.

**Comment by d-chambers** [comment]
Ok, I cleaned this up a bit by implementing a CD context manager. Also decided not to emit a custom message, pytest does a good job telling users what is wrong so we will just return its exit code.

**Comment by d-chambers** [comment]
pre-commit was complaining about some of the work done in previous PRs in docs so I cleaned that up too. 

@ThomasLecocq Do we need the cartopy import at the top of docs/conf.py?

**Comment by trichter** [comment]
> Do we need the cartopy import at the top of docs/conf.py?

Maybe this can be checked when dealing with #2956.

**Review by trichter — APPROVED** [review]

**Pull request merged**